katdal/chunkstore.py

```diff
@@ -77,32 +92,43 @@ def generate_chunks(shape, dtype, max_chunk_size, dims_to_split=None,
     """
     if dims_to_split is None:
         dims_to_split = range(len(shape))
-    dataset_size = np.prod(shape) * np.dtype(dtype).itemsize
-    # The ideal number of chunks to achieve requested chunk size (can be float)
-    num_chunks = dataset_size / max_chunk_size
-    # Start with the whole array as a single big chunk
-    chunks = [(s,) for s in shape]
+    if max_dim_elements is None:
+        max_dim_elements = {}
+
+    dim_elements = list(shape)
+    for i in dims_to_split:
+        if i in max_dim_elements and max_dim_elements[i] < shape[i]:
+            if power_of_two:
+                dim_elements[i] = _floor_power_of_two(max_dim_elements[i])
+            else:
+                dim_elements[i] = max_dim_elements[i]
+    cur_elements = int(np.prod(dim_elements))
```

Comment:
So you start with the maximum possible number of elements and work your way down to `max_elements`? :-) I think of `max_elements` more like `trg_elements`.